libs/api_common/registry_helper.cpp

```diff
@@ -0,0 +1,239 @@
+    } catch (...) {
+        result = ERROR_NOT_ENOUGH_MEMORY;
+    }
+
+    return result;
+}
+
+uint32_t
+write_registry_value_dword(_In_ const ebpf_registry_key_t* key, _In_z_ const wchar_t* value_name, uint32_t value)
+{
+    ebpf_assert(key->key);
+    return RegSetValueEx(key->key, value_name, 0, REG_DWORD, (PBYTE)&value, sizeof(value));
+}
+
+uint32_t
+create_registry_key(
+    _In_opt_ const ebpf_registry_key_t* root_key,
+    _In_ const wchar_t* sub_key,
+    uint32_t flags,
+    _Out_ ebpf_registry_key_t* key)
+{
+    key->key = nullptr;
+    if (root_key == nullptr) {
+        return ERROR_INVALID_PARAMETER;
+    }
+
+    return RegCreateKeyEx(root_key->key, sub_key, 0, nullptr, 0, flags, nullptr, &key->key, nullptr);
+}
+
+uint32_t
+open_registry_key(
+    _In_ const ebpf_registry_key_t* root_key,
+    _In_opt_z_ const wchar_t* sub_key,
+    uint32_t flags,
+    _Out_ ebpf_registry_key_t* key)
+{
+    ebpf_assert(root_key != nullptr);
+
+    return RegOpenKeyEx(root_key->key, sub_key, 0, flags, &key->key);
+}
+
+uint32_t
+delete_registry_key(_In_ const ebpf_registry_key_t* root_key, _In_z_ const wchar_t* sub_key)
+{
+    return RegDeleteKeyEx(root_key->key, sub_key, 0, 0);
+}
+
+uint32_t
+delete_registry_tree(_In_ const ebpf_registry_key_t* root_key, _In_opt_z_ const wchar_t* sub_key)
+{
+    return RegDeleteTree(root_key->key, sub_key);
+}
+
+uint32_t
+create_registry_key_ansi(
+    _In_ const ebpf_registry_key_t* root_key,
+    _In_z_ const char* sub_key,
+    uint32_t flags,
+    _Out_ ebpf_registry_key_t* key)
+{
+    uint32_t result;
+    try {
+        auto wide_string = _get_wstring_from_string(sub_key);
+        result = create_registry_key(root_key, wide_string.c_str(), flags, key);
+    } catch (...) {
+        result = ERROR_NOT_ENOUGH_MEMORY;
+    }
+
+    return result;
+}
+
+ebpf_result_t
+read_registry_value_string(HKEY key, _In_ const wchar_t* value_name, _Out_ wchar_t** value)
+{
+    uint32_t status = NO_ERROR;
+    DWORD type = REG_SZ;
+    DWORD value_size = 0;
+    wchar_t* string_value = nullptr;
+
+    *value = nullptr;
+    status = RegQueryValueEx(key, value_name, 0, &type, nullptr, &value_size);
+    if (status != ERROR_SUCCESS || type != REG_SZ) {
+        return win32_error_code_to_ebpf_result(status);
+    }
+
+    string_value = (wchar_t*)ebpf_allocate((value_size + sizeof(wchar_t)));
+    if (string_value == nullptr) {
+        status = ERROR_NOT_ENOUGH_MEMORY;
+        return win32_error_code_to_ebpf_result(status);
+    }
+
+    status = RegQueryValueEx(key, value_name, 0, &type, (PBYTE)string_value, &value_size);
+    if (status != ERROR_SUCCESS) {
+        goto Exit;
+    }
+    *value = string_value;
+    string_value = nullptr;
+
+Exit:
+    if (string_value) {
+        ebpf_free(string_value);
+    }
+    return win32_error_code_to_ebpf_result(status);
+}
+
+ebpf_result_t
+read_registry_value_dword(_In_ HKEY key, _In_ const wchar_t* value_name, _Out_ uint32_t* value)
```

Comment:
since no allocation required here, consider removing the wrapper.

Reply:
Added these wrappers for consistency with other APIs -- so that all registry APIs look similar. Also I still need to add ebpf store APIs to delete the entries from the store (in another PR), so these "read / delete" APIs will need wrappers so that calling code can execute both in user mode and kernel mode.